contrib/runners/inquirer/inquirer.py

```diff
@@ -0,0 +1,100 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import uuid
+
+from st2common import log as logging
+from st2common.constants.action import LIVEACTION_STATUS_SUCCEEDED
+from st2common.constants.action import LIVEACTION_STATUS_FAILED
+from st2common.constants.action import LIVEACTION_STATUS_PENDING
+from st2common.constants.triggers import INQUIRY_TRIGGER
+from st2common.models.system.common import ResourceReference
+from st2common.runners.base import ActionRunner
+from st2common.runners import python_action_wrapper
+from st2common.services import action as action_service
+from st2common.transport.reactor import TriggerDispatcher
+from st2common.util import action_db as action_utils
+
+LOG = logging.getLogger(__name__)
+
+__all__ = [
+    'get_runner',
+    'Inquirer',
+]
+
+# constants to lookup in runner_parameters.
+RUNNER_SCHEMA = 'schema'
+RUNNER_ROLES = 'roles'
+RUNNER_USERS = 'users'
+RUNNER_TAG = 'tag'
+
+#TODO
+BASE_DIR = os.path.dirname(os.path.abspath(python_action_wrapper.__file__))
+
+
+def get_runner():
+    'RunnerTestCase',
+    return Inquirer(str(uuid.uuid4()))
+
+
+class Inquirer(ActionRunner):
+    """This runner implements the ability to ask for more input during a workflow
+    """
+
+    def __init__(self, runner_id):
+        super(Inquirer, self).__init__(runner_id=runner_id)
+        self.trigger_dispatcher = TriggerDispatcher(LOG)
+
+    def pre_run(self):
+        super(Inquirer, self).pre_run()
+
+        # TODO :This is awful, but the way "runner_parameters" and other variables get
+        # assigned on the runner instance is even worse. Those arguments should
+        # be passed to the constructor.
+        self.schema = self.runner_parameters.get(RUNNER_SCHEMA, None)
+        self.roles_param = self.runner_parameters.get(RUNNER_ROLES, None)
+        self.users_param = self.runner_parameters.get(RUNNER_USERS, None)
+
+    def run(self, action_parameters):
+
+        # Retrieve existing result (initialize if needed)
+        liveaction_db = action_utils.get_liveaction_by_id(self.liveaction_id)
+        response_data = liveaction_db.result.get("response_data")
+        if not response_data:
+            response_data = {"response": {}}
+
+        # Request pause of parent execution
+        parent = liveaction_db.context.get("parent")
+        if parent:
+            # TODO get current user
+            action_service.request_pause(parent, "st2admin")
+        else:
+            LOG.error("Inquiries must be run within a workflow.")
+            return (LIVEACTION_STATUS_FAILED, response_data, None)
+
+        # Validate repsonse and return
+        if action_service.validate_response(self.schema, response_data):
```

Comment:
Yep, when response is posted to the API endpoint, the endpoint should validate the response first.  Since it's a specialized endpoint, and if the response passes validation, why can't we mark the action execution complete and save the result there from the endpoint? Why pass it back to the runner to be re-run?  Passing it back to the inquiry runner to re-run is fine, it'll work but it'll introduce inefficiency and a 2nd unnecessary validation.  So i'm trying to understand what's the value.

And say we do pass it back to the inquiry runner for re-run, where's the logic to resume the workflow?

I'm not concern that the validation code is shared.  We should because I think action result should be defined by schema and validated at some point and so putting the validation in some shared module is fine with me.